common/djangoapps/student/roles.py

```diff
@@ -83,15 +111,32 @@ def get_user_roles(cls, user):
 
 class RoleCache:
     """
-    A cache of the CourseAccessRoles held by a particular user
+    A cache of the CourseAccessRoles held by a particular user.
+    Internal data structures should be accessed by getter and setter methods;
+    don't use `_roles_by_course_id` or `_roles` directly.
+    _roles_by_course_id: This is the data structure as saved in the RequestCache.
+        It contains all roles for a user as a dict that's keyed by course_id.
+        The key ROLE_CACHE_UNGROUPED_COURSES_KEY is used for all roles
+        that are not associated with a course.
+    _roles: This is a set of all roles for a user, ungrouped. It's used for some types of
+        lookups and collected from _roles_by_course_id on initialization
+        so that it doesn't need to be recalculated.
+
     """
     def __init__(self, user):
         try:
-            self._roles = BulkRoleCache.get_user_roles(user)
+            self._roles_by_course_id = BulkRoleCache.get_user_roles(user)
         except KeyError:
-            self._roles = set(
-                CourseAccessRole.objects.filter(user=user).all()
-            )
+            self._roles_by_course_id = {}
+            roles = CourseAccessRole.objects.filter(user=user).all()
+            for role in roles:
+                course_id = get_role_cache_course_key(role.course_id)
+                if not self._roles_by_course_id.get(course_id):
+                    self._roles_by_course_id[course_id] = set()
+                self._roles_by_course_id[course_id].add(role)
+        self._roles = set()
+        for roles_for_course in self._roles_by_course_id.values():
+            self._roles.update(roles_for_course)
```

Comment:
Question: Shouldn't this kind of refresh logic be delegated to something in `BulkRoleCache`?

Reply:
Maybe it should. The BulkRoleCache is responsible for roles in general it seems and this is scoped to a specific user's roles, so you could maybe make some case for it to stay here.

The refresh happened here originally as well, so I tried to change as few things as I could to avoid breaking stuff. With your guidance here I could see about moving this logic to the BulkRoleCache in some way.

Generally I wasn't very impressed with the design of the RolesCache and BulkRoleCache, given that
- the performance was not optimal, as I've explained
- the separation between RolesCache and BulkRolesCache is not very clear since as you said, maybe this refresh logic belongs in the BulkRoleCache
- I'm not quite sure why you have a RequestCache with a BulkRoleCache but then you instead cache a single user's roles in the `_roles` property of the RoleCache here instead of the RequestCache. Is the RequestCache somehow slower, or why not just fetch the roles you want from the RequestCache and maybe avoid the whole problem we're having?
- the private data attribute `_roles` was accessed directly instead of by a getter in some external code which means that changing it means you may break places in the codebase you aren't aware of, which is why I tried to not change that attribute's internal data structure and why I was so concerned about changing any code here
- The roleCache is initialized and saved in a user's `_.roles` attribute so you'll need to do something like `user_roles = current_user._roles._roles` which is extremely confusing
- The BulkRoleCache's prefetch method uses very weird logic to set or overwrite the RequestCache that completely ignores the RequestCache's setter methods

I could probably find some more things that bother me.
Now a few things of these I'm attempting to fix here but I'm not trying to fix the overall design that shows the problems you see above. I'll try to keep these two things separate, which is why I probably wouldn't move the refresh logic here to the BulkRoleCache.

It probably would make sense to improve the architecture of this RoleCache in a separate issue, though. I'd love to see that.

Reply:
> I'm not quite sure why you have a RequestCache with a BulkRoleCache but then you instead cache a single user's roles in the _roles property of the RoleCache here instead of the RequestCache. Is the RequestCache somehow slower, or why not just fetch the roles you want from the RequestCache and maybe avoid the whole problem we're having?

I'd have to dig into the commit history, but keep in mind that things like this sometimes spring up as a series of bandaids that deal with specific bugs/issues over time, and don't necessarily add up to a coherent design intent. (Though sometimes they do, and those can be mind-bending. 😛 )

Reply:
> Outside of this implementation, it's anyway very weird how the RoleCache is randomly saved as an attribute a user has. The way it's done is just that some code somewhere suddenly adds this attribute _roles to a user object, which is not a set of roles but instead an instance of RoleCache. The problem besides the naming is that that's not a standard attribute a user has and you'll just go through code line by line to see if the code somewhere instantiates a RoleCache and adds it as such an attribute to a user. It's completely intransparent whether the user object will have an attached RoleCache or not at a given moment in time

I think this is part of the early days of fumbling around as to how to cache something at the request level. There's no real built-in thing for this in Django, but there _is_ a User object that's passed into the view as part of the Request, and that Request can be munged with in middleware. So if you want to stick some data onto a user for the duration of the request, you can slap it in there. Not something I'd recommend as a good practice, but I'd bet that's where it started from.